synapse/push/baserules.py

```diff
@@ -14,433 +14,544 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-import copy
-from typing import Any, Dict, List
+import itertools
+from typing import Dict, Iterator, List, Mapping, Sequence, Tuple, Union
 
-from synapse.push.rulekinds import PRIORITY_CLASS_INVERSE_MAP, PRIORITY_CLASS_MAP
+import attr
 
+from synapse.config.experimental import ExperimentalConfig
+from synapse.push.rulekinds import PRIORITY_CLASS_MAP
 
-def list_with_base_rules(rawrules: List[Dict[str, Any]]) -> List[Dict[str, Any]]:
-    """Combine the list of rules set by the user with the default push rules
 
-    Args:
-        rawrules: The rules the user has modified or set.
+@attr.s(auto_attribs=True, slots=True, frozen=True)
+class PushRule:
+    rule_id: str
+    priority_class: int
+    conditions: Sequence[Mapping[str, str]]
+    actions: Sequence[Union[str, Mapping]]
+    default: bool = False
+    default_enabled: bool = True
 
-    Returns:
-        A new list with the rules set by the user combined with the defaults.
+
+@attr.s(auto_attribs=True, slots=True, frozen=True, weakref_slot=False)
```

Comment:
Why the weakref here?

Reply:
Attrs automatically adds a slot for weakrefs, which we don't use so I idly turned it off :shrug: